### PR TITLE
Metrics: show top 5 services disk IO

### PR DESCRIPTION
**Pull request by tomasmatus**
Currently running a test if something in our CI has this cgroup controller enabled.

### CODE REVIEW

File: pkg/metrics/metrics.jsx [hunk 1]
```diff
@@ -520,21 +525,46 @@
...
+                            pV[i] = [top5IO[i][0], top5IO[i][2], top5IO[i][3], top5IO[i][4], cgroup[1], cgroup[2]];
+                            return pV;
+                        }
+                    }
+
+                    return pV;
+                }, []);
+
+        newState.topServicesDiskIO = topServicesDiskIO.map(
+            ([key, is_user, is_container, userid, read, write]) => cgroupRow(key, is_user, is_container, userid, cockpit.format_bytes_per_sec(read), cockpit.format_bytes_per_sec(write))
```
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:567]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18070-20221220-124215-d2339978-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx [hunk 2]
```diff
@@ -520,21 +525,46 @@ class CurrentMetrics extends React.Component {
...
+                .reduce((pV, cgroup) => {
+                    for (let i = 0; i < top5IO.length; i++) {
+                        // is container || cgroup name match
+                        if ((cgroup[0].match(podmanCgroupRe) && cgroup[0].includes(top5IO[i][0])) ||
+                            (top5IO[i][4] === cgroup[0])) {
+                            // push [key, is_user, is_container, userid, read, write]
+                            pV[i] = [top5IO[i][0], top5IO[i][2], top5IO[i][3], top5IO[i][4], cgroup[1], cgroup[2]];
+                            return pV;
+                        }
+                    }
```
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:561]: These 9 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18070-20221221-101537-567d69a1-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:567]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18070-20221221-101537-567d69a1-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx [hunk 3]
```diff
@@ -815,6 +845,18 @@
...
+                        { this.state.topServicesDiskIO.length > 0 &&
+                            <Table
+                                variant={TableVariant.compact}
+                                gridBreakPoint={TableGridBreakpoint.none}
+                                borders={false}
+                                aria-label={ _("Top 5 services disk IO") }
+                                cells={ [{ title: _("Service"), transforms: [cellWidth(60)] }, _("Read"), _("Write")] }
+                                rows={this.state.topServicesDiskIO}>
+                                <TableHeader />
+                                <TableBody />
```
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:858]: These 9 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18070-20221221-101537-567d69a1-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:561]: These 9 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18070-20221222-100304-73185077-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:567]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18070-20221222-100304-73185077-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/metrics/metrics.jsx (hunk 3 above)
Comment (cockpituous) [inline pkg/metrics/metrics.jsx:858]: These 9 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18070-20221222-100304-73185077-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
Some things discussed in our meeting today:

* PCP collects these metrics as well but using a kernel module? Is that something to look into?
* IOTop is a userspace alternative:

image

Doesn't feel too efficient.

**Comment by tomasmatus** [comment]
@jelly so I probably mixed this a bit with network related info, for storage pcp doesn't use submodule but it uses "storage accounting" from kernel (if it's enabled)

from `pcp-atop` manpage:
```
  d    Show disk-related output.

            When "storage accounting" is active in the kernel, the following  fields  are  shown:
            process-id,  amount of data read from disk, amount of data written to disk, amount of
            data that was  written  but  has  been  withdrawn  again  (WCANCL),  disk  occupation
            percentage and process name.
```

**Comment by jelly** [comment]
> @jelly so I probably mixed this a bit with network related info, for storage pcp doesn't use submodule but it uses "storage accounting" from kernel (if it's enabled)
> 
> from `pcp-atop` manpage:
> 
> ```
>   d    Show disk-related output.
> 
>             When "storage accounting" is active in the kernel, the following  fields  are  shown:
>             process-id,  amount of data read from disk, amount of data written to disk, amount of
>             data that was  written  but  has  been  withdrawn  again  (WCANCL),  disk  occupation
>             percentage and process name.
> ```

This looks like the description from the man page of atop let's read what it does.

**Comment by jelly** [comment]
atop just reads https://github.com/Atoptool/atop/blob/8365a7583e08a27ed14aedd2d0909dbde0513f18/photoproc.c#L61

**Comment by marusak** [comment]
@tomasmatus this one can be closed, right?

**Comment by tomasmatus** [comment]
closed in favor of #17692

**Pull request closed without merging**